### ISSUE TITLE
Network misbehavior

### ISSUE THREAD

**Issue by palexster** [liqotech/liqo#314]
After Liqo installation, Kindnet crashes stating it is unable to reconcile routes. This is the function that seems to crash dealing with Liqonet: https://github.com/kubernetes-sigs/kind/blob/d7f948dd8c00084d6ee30eb953471ce3ce375455/images/kindnetd/cmd/kindnetd/main.go#L137

KindNet Logs:
```
I1004 11:25:59.390304       1 main.go:65] hostIP = 172.18.0.6
podIP = 172.18.0.6
I1004 11:25:59.390670       1 main.go:74] setting mtu 1500 for CNI 
I1004 11:25:59.778503       1 main.go:168] Handling node with IP: 10.200.1.14
I1004 11:25:59.778531       1 main.go:169] Node liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR 10.200.3.0/24 
I1004 11:25:59.778830       1 main.go:124] Failed to reconcile routes, retrying after error: network is unreachable
I1004 11:25:59.778843       1 main.go:168] Handling node with IP: 10.200.1.14
I1004 11:25:59.778849       1 main.go:169] Node liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR 10.200.3.0/24 
I1004 11:25:59.778936       1 main.go:124] Failed to reconcile routes, retrying after error: network is unreachable
I1004 11:26:00.779140       1 main.go:168] Handling node with IP: 10.200.1.14
I1004 11:26:00.779363       1 main.go:169] Node liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR 10.200.3.0/24 
I1004 11:26:00.779707       1 main.go:124] Failed to reconcile routes, retrying after error: network is unreachable
I1004 11:26:02.779986       1 main.go:168] Handling node with IP: 10.200.1.14
I1004 11:26:02.780022       1 main.go:169] Node liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR 10.200.3.0/24 
I1004 11:26:02.780227       1 main.go:124] Failed to reconcile routes, retrying after error: network is unreachable
I1004 11:26:05.780436       1 main.go:168] Handling node with IP: 10.200.1.14
I1004 11:26:05.780464       1 main.go:169] Node liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR 10.200.3.0/24 
I1004 11:26:05.780638       1 main.go:124] Failed to reconcile routes, retrying after error: network is unreachable
panic: Maximum retries reconciling node routes: network is unreachable

goroutine 1 [running]:
main.main()
	/go/src/cmd/kindnetd/main.go:128 +0x893
```

Ip Route:
```
default via 172.18.0.1 dev eth0 
10.75.0.0/16 via 192.168.200.7 dev liqonet 
10.141.0.0/16 via 192.168.200.7 dev liqonet 
10.200.0.2 dev vethc8f0b0ee scope host 
10.200.0.3 dev veth7ad31e05 scope host 
10.200.0.4 dev veth6292c0c6 scope host 
10.200.1.0/24 via 172.18.0.7 dev eth0 
172.18.0.0/16 dev eth0 proto kernel scope link src 172.18.0.6 
192.168.200.0/24 dev liqonet proto kernel scope link src 192.168.200.6 
```

Iptables:
```
root@liqo-cluster1-control-plane:/# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
LIQO-INPUT  udp  --  anywhere             anywhere             udp
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-EXTERNAL-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes externally-visible service portals */
KUBE-FIREWALL  all  --  anywhere             anywhere            

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
LIQO-FORWARD  all  --  anywhere             anywhere            
KUBE-FORWARD  all  --  anywhere             anywhere             /* kubernetes forwarding rules */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-FIREWALL  all  --  anywhere             anywhere            

Chain KUBE-EXTERNAL-SERVICES (1 references)
target     prot opt source               destination         

Chain KUBE-FIREWALL (2 references)
target     prot opt source               destination         
DROP       all  --  anywhere             anywhere             /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
DROP       all  -- !127.0.0.0/8          127.0.0.0/8          /* block incoming localnet connections */ ! ctstate RELATED,ESTABLISHED,DNAT

Chain KUBE-FORWARD (1 references)
target     prot opt source               destination         
DROP       all  --  anywhere             anywhere             ctstate INVALID
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding rules */ mark match 0x4000/0x4000
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED

Chain KUBE-KUBELET-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-PROXY-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-SERVICES (3 references)
target     prot opt source               destination         

Chain LIQO-FORWARD (1 references)
target     prot opt source               destination         
LIQO-FRWD-CLS-f38be2c9  all  --  anywhere             10.141.0.0/16       
LIQO-FRWD-CLS-3dd6f45d  all  --  anywhere             10.75.0.0/16        

Chain LIQO-FRWD-CLS-3dd6f45d (1 references)
target     prot opt source               destination         
ACCEPT     all  --  anywhere             10.75.0.0/16        

Chain LIQO-FRWD-CLS-f38be2c9 (1 references)
target     prot opt source               destination         
ACCEPT     all  --  anywhere             10.141.0.0/16       

Chain LIQO-INPT-CLS-3dd6f45d (1 references)
target     prot opt source               destination         
ACCEPT     all  --  10.200.0.0/16        10.75.0.0/16        

Chain LIQO-INPT-CLS-f38be2c9 (1 references)
target     prot opt source               destination         
ACCEPT     all  --  10.200.0.0/16        10.141.0.0/16       

Chain LIQO-INPUT (1 references)
target     prot opt source               destination         
ACCEPT     udp  --  anywhere             anywhere             udp dpt:4789
LIQO-INPT-CLS-f38be2c9  all  --  anywhere             10.141.0.0/16       
LIQO-INPT-CLS-3dd6f45d  all  --  anywhere             10.75.0.0/16  
```

**Comment by alacuku** [comment]
Well, this bug does not seem to be related to liqonet but how the kindnet manages the routes between the nodes. It expects the nodes to be in the same L3 network. Based on the logs the `PodCIDR` is 10.200.0.0/16. It is divided in subnets with netmask 255.255.255.0. Each new subnet is allocated to a node. Each pod deployed an on a given node takes an IP from the subnet assigned to the same node. In order for the pods living in different nodes to communicate between them KindNet configures the routing table. Being the nodes on the same L3 network every thing work smothly. 
When we create the virtual node in this case `liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5 has CIDR` with subnet 10.200.3.0/24 taken by the PodCIDR 10.200.0.0/16 the KindNet running on the other nodes wants to make this subnet reachable by configuring a route that sent the traffic to the virtual node. The `liqo-3dd6f45d-b53a-4f67-8416-505c4d5ddba5` is a pod, running on a node, and to it is given an IP taken by the `PodCIDR`. From the logs we see that the virtual node has  IP 10.200.1.14. Looking at the source code of KindNet (https://github.com/kubernetes-sigs/kind/blob/d7f948dd8c00084d6ee30eb953471ce3ce375455/images/kindnetd/cmd/kindnetd/routes.go#L36) it is trying to insert a route for the subnet 10.200.3.0/24 using as gateway the node IP that is 10.200.1.14. It does not succeeds because the virtual node is not directly reachable from the nodes except for the node where the deployment of the virtualkubelet is running.
To solve this problem for KindNet it would be enough to remove the `PodCIDR` from the node specs of the virtual node.

**Comment by palexster** [comment]
Thanks a lot @alacuku. I implemented a simple patch as you suggested (#311) and it worked fine: https://github.com/liqotech/liqops/actions/runs/288131231. It has to be checked if this does not break compatibility with other CNIs.

**Comment by NominalTrajectory** [comment]
Hi there! I think I'm experiencing the exact same issue today testing Liqo. 

I'm following a quickstart guide described here https://doc.liqo.io/gettingstarted/helloworld/ and using the exact same configurations provided using kind v0.11.1 go1.16.4 darwin/amd64 with Docker Desktop on Mac.

Liqo installs successfully, but as soon as the virtual-kubelet gets created, I see crashing kindnet and kube-proxy pods.

![Screenshot 2022-02-04 at 13 37 05](https://user-images.githubusercontent.com/47399749/152530277-d8b5ef45-b787-4ac4-968c-535d92d39794.png)

My kindnet logs are:
```
I0204 12:36:30.478346       1 main.go:316] probe TCP address cluster1-control-plane:6443
I0204 12:36:30.480330       1 main.go:102] connected to apiserver: https://cluster1-control-plane:6443
I0204 12:36:30.480369       1 main.go:107] hostIP = 172.19.0.3
podIP = 172.19.0.3
I0204 12:36:30.480601       1 main.go:116] setting mtu 1500 for CNI
I0204 12:36:30.480640       1 main.go:146] kindnetd IP family: "ipv4"
I0204 12:36:30.480666       1 main.go:150] noMask IPv4 subnets: [10.200.0.0/16]
I0204 12:36:31.075989       1 main.go:223] Handling node with IPs: map[172.19.0.3:{}]
I0204 12:36:31.076049       1 main.go:227] handling current node
I0204 12:36:31.078393       1 main.go:223] Handling node with IPs: map[172.19.0.2:{}]
I0204 12:36:31.078506       1 main.go:250] Node cluster1-worker has CIDR [10.200.1.0/24]
I0204 12:36:31.078776       1 main.go:223] Handling node with IPs: map[10.200.1.9:{}]
I0204 12:36:31.078816       1 main.go:250] Node liqo-b761b1b3-d06c-46a3-94a0-a025c0d2f147 has CIDR [10.200.2.0/24]
I0204 12:36:31.078941       1 routes.go:46] Adding route {Ifindex: 0 Dst: 10.200.2.0/24 Src: <nil> Gw: 10.200.1.9 Flags: [] Table: 0}
I0204 12:36:31.079222       1 main.go:204] Failed to reconcile routes, retrying after error: network is unreachable
I0204 12:36:31.079427       1 main.go:223] Handling node with IPs: map[172.19.0.3:{}]
I0204 12:36:31.079493       1 main.go:227] handling current node
I0204 12:36:31.079511       1 main.go:223] Handling node with IPs: map[172.19.0.2:{}]
I0204 12:36:31.079762       1 main.go:250] Node cluster1-worker has CIDR [10.200.1.0/24]
I0204 12:36:31.080011       1 main.go:223] Handling node with IPs: map[10.200.1.9:{}]
I0204 12:36:31.080094       1 main.go:250] Node liqo-b761b1b3-d06c-46a3-94a0-a025c0d2f147 has CIDR [10.200.2.0/24]
I0204 12:36:31.080244       1 routes.go:46] Adding route {Ifindex: 0 Dst: 10.200.2.0/24 Src: <nil> Gw: 10.200.1.9 Flags: [] Table: 0}
I0204 12:36:31.080403       1 main.go:204] Failed to reconcile routes, retrying after error: network is unreachable
I0204 12:36:32.080503       1 main.go:223] Handling node with IPs: map[172.19.0.3:{}]
I0204 12:36:32.080603       1 main.go:227] handling current node
I0204 12:36:32.080628       1 main.go:223] Handling node with IPs: map[172.19.0.2:{}]
I0204 12:36:32.080651       1 main.go:250] Node cluster1-worker has CIDR [10.200.1.0/24]
I0204 12:36:32.080802       1 main.go:223] Handling node with IPs: map[10.200.1.9:{}]
I0204 12:36:32.080827       1 main.go:250] Node liqo-b761b1b3-d06c-46a3-94a0-a025c0d2f147 has CIDR [10.200.2.0/24]
I0204 12:36:32.081020       1 routes.go:46] Adding route {Ifindex: 0 Dst: 10.200.2.0/24 Src: <nil> Gw: 10.200.1.9 Flags: [] Table: 0}
I0204 12:36:32.081097       1 main.go:204] Failed to reconcile routes, retrying after error: network is unreachable
I0204 12:36:34.081567       1 main.go:223] Handling node with IPs: map[172.19.0.3:{}]
I0204 12:36:34.081681       1 main.go:227] handling current node
I0204 12:36:34.081715       1 main.go:223] Handling node with IPs: map[172.19.0.2:{}]
I0204 12:36:34.081740       1 main.go:250] Node cluster1-worker has CIDR [10.200.1.0/24]
I0204 12:36:34.081976       1 main.go:223] Handling node with IPs: map[10.200.1.9:{}]
I0204 12:36:34.082001       1 main.go:250] Node liqo-b761b1b3-d06c-46a3-94a0-a025c0d2f147 has CIDR [10.200.2.0/24]
I0204 12:36:34.082069       1 routes.go:46] Adding route {Ifindex: 0 Dst: 10.200.2.0/24 Src: <nil> Gw: 10.200.1.9 Flags: [] Table: 0}
I0204 12:36:34.082150       1 main.go:204] Failed to reconcile routes, retrying after error: network is unreachable
I0204 12:36:37.082937       1 main.go:223] Handling node with IPs: map[172.19.0.3:{}]
I0204 12:36:37.083409       1 main.go:227] handling current node
I0204 12:36:37.083493       1 main.go:223] Handling node with IPs: map[172.19.0.2:{}]
I0204 12:36:37.083546       1 main.go:250] Node cluster1-worker has CIDR [10.200.1.0/24]
I0204 12:36:37.083698       1 main.go:223] Handling node with IPs: map[10.200.1.9:{}]
I0204 12:36:37.083771       1 main.go:250] Node liqo-b761b1b3-d06c-46a3-94a0-a025c0d2f147 has CIDR [10.200.2.0/24]
I0204 12:36:37.083964       1 routes.go:46] Adding route {Ifindex: 0 Dst: 10.200.2.0/24 Src: <nil> Gw: 10.200.1.9 Flags: [] Table: 0}
I0204 12:36:37.084140       1 main.go:204] Failed to reconcile routes, retrying after error: network is unreachable
panic: Maximum retries reconciling node routes: network is unreachable

goroutine 1 [running]:
main.main()
	/go/src/cmd/kindnetd/main.go:208 +0x1327
```



@palexster can this issue be reopened?
